### PR TITLE
Telemetry UI and final touches

### PR DESCRIPTION
**Pull request by bunnei**
This is the fourth part of a set of changes to add a telemetry framework to Citra. In the near term, this will be used for user-submitted compatibility data to our own custom service endpoint (services.citra-emu.org). My goal, however, is to make this very generic, and allow us to easily log useful/interesting data fields throughout the emulator, and then upload them to our service at a later time. In the long term, we may use this for anonymous telemetry, critical errors and crash reports, measuring how often features are used, etc.

With this change, I've added the following:
* Settings UI for configuring telemetry related fields
* "Callouts", one-time important messages to the user, which is currently used to indicate the addition of telemetry
* TelemetryId, a user-unique random number used for grouping telemetry data by user
* Support for unauthenticated/anonymous telemetry
* Logging of the emulation frontend type (Qt or SDL)

As I mentioned above, this is a multi-part submission. Here is what has been completed thus far, as well as what is to come next:

- [x] [#2683] Common interface for submitting data fields to the telemetry log
- [x] [#2683] Dummy backend that does nothing with the telemetry log
- [x] [#2683] Core instantiation of the telemetry interface with a single global log used for game compatibility
- [x] [#2683] A few example fields are logged in core
- [x] [#2819] Implement a backend that does an authenticated log submission to our service
- [x] [#2823] Implement remaining fields for game compatibility submission
- [x] Implement Qt UI for authenticating users and user-submitted data
- [ ] Support multiple logs for different purposes

Obligatory UI screenshots:

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/2897)

### CODE REVIEW

**Comment by mailwl** [comment]
omg! so long password

File: src/web_service/web_backend.cpp [hunk 1]
```diff
@@ -5,48 +5,37 @@
...
 static constexpr char API_VERSION[]{"1"};
-static constexpr char ENV_VAR_USERNAME[]{"CITRA_WEB_SERVICES_USERNAME"};
-static constexpr char ENV_VAR_TOKEN[]{"CITRA_WEB_SERVICES_TOKEN"};
 
-static std::string GetEnvironmentVariable(const char* name) {
-    const char* value{getenv(name)};
-    if (value) {
-        return value;
+void PostJson(const std::string& url, const std::string& data) {
+    if (!Settings::values.enable_telemetry) {
```
Comment (B3n30) [inline src/web_service/web_backend.cpp]: This check should be in TelemetryJson::Complete() since web_backend isn't for telemetry only, right?

File: src/web_service/web_backend.cpp [hunk 2]
```diff
@@ -5,48 +5,37 @@
...
 
 static constexpr char API_VERSION[]{"1"};
-static constexpr char ENV_VAR_USERNAME[]{"CITRA_WEB_SERVICES_USERNAME"};
-static constexpr char ENV_VAR_TOKEN[]{"CITRA_WEB_SERVICES_TOKEN"};
 
-static std::string GetEnvironmentVariable(const char* name) {
-    const char* value{getenv(name)};
-    if (value) {
-        return value;
+void PostJson(const std::string& url, const std::string& data) {
```
Comment (B3n30) [inline src/web_service/web_backend.cpp]: I think the web_service calling PostJson should also send a flag if the data should be sent anonymous if no lgin data is available, e.g.
`void PostJson(const std::string& url, const std::string& data, const bool enable_anonymous=true)`

File: src/web_service/web_backend.cpp (hunk 1 above)
Comment (bunnei) [inline src/web_service/web_backend.cpp]: Derp - yes

File: src/web_service/web_backend.cpp (hunk 2 above)
Comment (bunnei) [inline src/web_service/web_backend.cpp]: Yes - this will make it easier for other callers. I wasn't thinking about you guys when I wrote it :)

**Comment by jroweboy** [comment]
Calling dibs on review, but i might not be able to review it till next week since i'm outta town this weekend. sorry!

File: src/web_service/web_backend.cpp [hunk 3]
```diff
@@ -5,48 +5,37 @@
 #include <cpr/cpr.h>
 #include <stdlib.h>
 #include "common/logging/log.h"
+#include "core/settings.h"
```
Comment (bunnei) [inline src/web_service/web_backend.cpp]: This is not new with this PR, but will fix...

File: src/core/telemetry_session.cpp [hunk 4]
```diff
@@ -3,8 +3,10 @@
 // Refer to the license.txt file included.
 
 #include <cstring>
+#include <cryptopp/osrng.h>
```
Comment (bunnei) [inline src/core/telemetry_session.cpp]: Huh? This is already used in Core..?

Comment (wwylele) [inline src/core/telemetry_session.cpp]: oops nvm, was looking at the wrong file :p

Comment (wwylele) [inline src/core/telemetry_session.cpp]: oops nvm, was looking at the wrong file :p

File: src/citra_qt/main.cpp [hunk 5]
```diff
@@ -48,6 +48,47 @@
...
+    msg.setText(message);
+    msg.setStandardButtons(QMessageBox::Ok);
+    msg.setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+    msg.setStyleSheet("QLabel{min-width: 900px;}");
+    msg.exec();
+}
+
+void GMainWindow::ShowCallouts() {
+    static const QString telemetry_message =
+        tr("To improve Citra, we have begun collecting anonymous usage data. No private or "
```
Comment (danzel) [inline src/citra_qt/main.cpp]: Are we planning to come back and change this text again when the feature isn't new?
Maybe "To improve Citra, we collect anonymous usage data." instead?

File: src/web_service/web_backend.cpp [hunk 6]
```diff
@@ -10,43 +10,33 @@
...
         return;
     }
 
-    cpr::PostAsync(cpr::Url{url}, cpr::Body{data}, cpr::Header{{"Content-Type", "application/json"},
-                                                               {"x-username", GetUsername()},
-                                                               {"x-token", GetToken()},
-                                                               {"api-version", API_VERSION}});
+    if (are_credentials_provided) {
+        // Authenticated request if credentials are provided
+        cpr::PostAsync(cpr::Url{url}, cpr::Body{data},
```
Comment (wwylele) [inline src/web_service/web_backend.cpp]: I found that `PostAsync` doesn't actually go async when I tested it. The reason is that the function returns a future, which we doesn't hold. The future gets destructed immediately after the function call, and it gets blocked in its destructor to wait for the Post to finish.

Probably we can do `std::thread(cpr::Post, ...).detach()` instead

File: src/citra_qt/main.cpp (hunk 5 above)
Comment (bunnei) [inline src/citra_qt/main.cpp]: Sure, I will just make the above change now

File: src/citra_qt/configuration/configure_web.ui [hunk 7]
```diff
@@ -0,0 +1,153 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (bunnei) [reply src/citra_qt/configuration/configure_web.ui]: addressed

**Comment by bunnei** [comment]
@B3n30 while that would probably be slightly better, I wasn't sure how to do that and it's not really important. Disabling web_services won't be an officially supported configuration, it's just for faster dev builds and unofficial stuff. For these scenarios, the extra tab doesn't really harm anything.

File: src/web_service/web_backend.cpp [hunk 8]
```diff
@@ -2,15 +2,31 @@
 // Licensed under GPLv2 or any later version
 // Refer to the license.txt file included.
 
+#include <thread>
 #include <cpr/cpr.h>
 #include <stdlib.h>
+
```
Comment (wwylele) [inline src/web_service/web_backend.cpp]: no need this new line
